src/shogun/evaluation/CrossValidatedCalibration.cpp

```diff
@@ -0,0 +1,406 @@
+
+		SG_DEBUG("finished evaluation\n")
+		features->remove_subset();
+
+		/* clean up, remove subsets */
+		labels->remove_subset();
+		if (get_global_parallel()->get_num_threads() != 1)
+		{
+			SG_UNREF(machine);
+			SG_UNREF(features);
+			SG_UNREF(labels);
+		}
+	}
+
+	SG_DEBUG("done unlocked evaluation\n", get_name())
+	return true;
+}
+
+bool CCrossValidatedCalibration::train_locked(SGVector<index_t> indices)
+{
+
+	index_t num_subsets = m_splitting_strategy->get_num_subsets();
+
+	m_calibration_machines = new CDynamicObjectArray(num_subsets);
+
+	SG_DEBUG(
+	    "building index sets for %d-fold cross-validated calibration\n",
+	    num_subsets)
+
+	/* build index sets */
+	m_splitting_strategy->build_subsets();
+
+	if (m_machine->is_data_locked())
+	{
+		SG_ERROR("please lock the data before running train_locked")
+		return false;
+	}
+
+	SG_DEBUG("starting unlocked evaluation\n", get_name())
+	/* tell machine to store model internally
+	 * (otherwise changing subset of features will kaboom the classifier) */
+	m_machine->set_store_model_features(true);
+
+	for (index_t i = 0; i < num_subsets; ++i)
+	{
+		/* index subset for training, will be freed below */
+		SGVector<index_t> inverse_subset_indices =
+		    m_splitting_strategy->generate_subset_inverse(i);
+
+		/* train machine on training features */
+		m_machine->train_locked(inverse_subset_indices);
+
+		/* feature subset for calibration */
+		SGVector<index_t> subset_indices =
+		    m_splitting_strategy->generate_subset_indices(i);
+
+		/* set subset for testing labels */
+		m_labels->add_subset(subset_indices);
+
+		/* produce output for desired indices */
+		CCalibration* calibrator = new CCalibration();
+		calibrator->set_machine((CMachine*)m_machine->clone());
+		calibrator->set_labels((CBinaryLabels*)m_labels->clone());
+		calibrator->set_calibration_method(
+		    (CCalibrationMethod*)m_calibration_method->clone());
+
+		bool trained = calibrator->train_locked(subset_indices);
+
+		if (!trained)
+		{
+			return false;
+		}
+
+		m_calibration_machines->set_element(calibrator, i);
+
+		/* remove subset to prevent side effects */
+		m_labels->remove_subset();
+
+		SG_DEBUG("done locked evaluation\n", get_name())
+	}
+	return true;
+}
+
+CBinaryLabels* CCrossValidatedCalibration::apply_binary(CFeatures* features)
+{
+	index_t num_machines = m_calibration_machines->get_num_elements();
+
+	CBinaryLabels* temp_result;
+	CMachine* temp_machine;
+	SGVector<float64_t> result_values;
+	CBinaryLabels* result_labels;
+
+	for (index_t i = 0; i < num_machines; ++i)
+	{
+		temp_machine = (CMachine*)m_calibration_machines->get_element(i);
+		temp_result = (CBinaryLabels*)temp_machine->apply(features);
+		if (i == 0)
+		{
+			result_values = temp_result->get_values();
+			result_labels = temp_result;
+		}
+		else
+		{
+			result_values += temp_result->get_values();
+		}
+	}
+
+#pragma omp parallel for
+	for (index_t i = 0; i < result_values.vlen; ++i)
```

Comment:
this could be substituted by a simple `linalg::scale`
has this code been borrowed from somewhere?

Reply:
I don't think that I borrowed this from somewhere. I'm not sure though..

Reply:
ok so then just change it to linalg::scale
the reason i've asked is because if it borrowed then we should change it there as well :)